File: stdlib/host/os/filesystem.go
package oshost

import (
	"context"
	"errors"
	"fmt"
	"io"
	"io/fs"
	"sort"
	"sync"
	"time"

	"github.com/d7z-team/mini-go/rpc"
)

// Filesystem is the host-neutral storage contract used by the os package.
type Filesystem interface {
	Open(name string, flag int, perm fs.FileMode) (File, error)
	Stat(name string) (fs.FileInfo, error)
	Lstat(name string) (fs.FileInfo, error)
	ReadDir(name string) ([]fs.DirEntry, error)
	Mkdir(name string, perm fs.FileMode) error
	Remove(name string) error
	Rename(oldName, newName string) error
	Readlink(name string) (string, error)
	Chtimes(name string, atime, mtime time.Time) error
	Getwd() (string, error)
	UserCacheDir() (string, error)
}

// File is the seekable directory-aware file shape required by the host adapter.
type File interface {
	io.Reader
	io.Writer
	io.ReaderAt
	io.WriterAt
	io.Seeker
	io.Closer
	Stat() (fs.FileInfo, error)
	ReadDir(count int) ([]fs.DirEntry, error)
}

type filesystemProviderBinding struct{ filesystemBackend Filesystem }

// NewFilesystemProvider exposes filesystem through the generated os RPC contract. The
// route owns exported file resources; the backend remains application-owned.
func NewFilesystemProvider(filesystemBackend Filesystem) (rpc.Provider, error) {
	if filesystemBackend == nil {
		return nil, errors.New("filesystem is required")
	}
	return NewOsFilesystemProvider(&filesystemProviderBinding{filesystemBackend: filesystemBackend})
}

func (binding *filesystemProviderBinding) Open(_ context.Context, name string, flag int64, perm uint32) (OsFileHandleHandler, OsOperationFault, error) {
	nativeFlag := int(flag)
	if int64(nativeFlag) != flag {
		return nil, operationFault(fs.ErrInvalid), nil
	}
	file, openErr := binding.filesystemBackend.Open(name, nativeFlag, fs.FileMode(perm))
	if openErr != nil {
		return nil, operationFault(openErr), nil
	}
	return &fileBinding{file: file}, OsOperationFault{}, nil
}

func (binding *filesystemProviderBinding) Stat(_ context.Context, name string, follow bool) (OsFileMetadata, OsOperationFault, error) {
	var info fs.FileInfo
	var err error
	if follow {
		info, err = binding.filesystemBackend.Stat(name)
	} else {
		info, err = binding.filesystemBackend.Lstat(name)
	}
	metadata, fault := fileInfoResult(info, err)
	return metadata, fault, nil
}

func (binding *filesystemProviderBinding) ReadDir(_ context.Context, name string) ([]OsFileMetadata, OsOperationFault, error) {
	entries, err := binding.filesystemBackend.ReadDir(name)
	if err != nil {
		return nil, operationFault(err), nil
	}
	sort.Slice(entries, func(i, j int) bool { return entries[i].Name() < entries[j].Name() })
	metadata, err := directoryEntries(entries)
	return metadata, operationFault(err), nil
}

func (binding *filesystemProviderBinding) Mkdir(_ context.Context, name string, perm uint32) (OsOperationFault, error) {
	return operationFault(binding.filesystemBackend.Mkdir(name, fs.FileMode(perm))), nil
}

func (binding *filesystemProviderBinding) Remove(_ context.Context, name string) (OsOperationFault, error) {
	return operationFault(binding.filesystemBackend.Remove(name)), nil
}

func (binding *filesystemProviderBinding) Rename(_ context.Context, oldName, newName string) (OsOperationFault, error) {
	return operationFault(binding.filesystemBackend.Rename(oldName, newName)), nil
}

func (binding *filesystemProviderBinding) Readlink(_ context.Context, name string) (string, OsOperationFault, error) {
	value, err := binding.filesystemBackend.Readlink(name)
	return value, operationFault(err), nil
}

func (binding *filesystemProviderBinding) Chtimes(_ context.Context, name string, atimeSeconds, atimeNanoseconds, mtimeSeconds, mtimeNanoseconds int64) (OsOperationFault, error) {
	err := binding.filesystemBackend.Chtimes(name, time.Unix(atimeSeconds, atimeNanoseconds), time.Unix(mtimeSeconds, mtimeNanoseconds))
	return operationFault(err), nil
}

func (binding *filesystemProviderBinding) Getwd(context.Context) (string, OsOperationFault, error) {
	value, err := binding.filesystemBackend.Getwd()
	return value, operationFault(err), nil
}

func (binding *filesystemProviderBinding) UserCacheDir(context.Context) (string, OsOperationFault, error) {
	value, err := binding.filesystemBackend.UserCacheDir()
	return value, operationFault(err), nil
}

type fileBinding struct {
	mu   sync.Mutex
	file File
}

func (binding *fileBinding) Read(_ context.Context, size int64) ([]uint8, OsOperationFault, error) {
	binding.mu.Lock()
	defer binding.mu.Unlock()
	if binding.file == nil {
		return nil, operationFault(fs.ErrClosed), nil
	}
	if size < 0 || size > 64<<20 {
		return nil, operationFault(fs.ErrInvalid), nil
	}
	data := make([]byte, int(size))
	count, err := binding.file.Read(data)
	data = data[:count]
	if errors.Is(err, io.EOF) {
		err = nil
	}
	return data, operationFault(err), nil
}

func (binding *fileBinding) Write(_ context.Context, data []uint8) (int64, OsOperationFault, error) {
	binding.mu.Lock()
	defer binding.mu.Unlock()
	if binding.file == nil {
		return 0, operationFault(fs.ErrClosed), nil
	}
	count, err := binding.file.Write(data)
	return int64(count), operationFault(err), nil
}

func (binding *fileBinding) ReadAt(_ context.Context, size, offset int64) ([]uint8, OsOperationFault, error) {
	binding.mu.Lock()
	defer binding.mu.Unlock()
	if binding.file == nil {
		return nil, operationFault(fs.ErrClosed), nil
	}
	if size < 0 || size > 64<<20 || offset < 0 {
		return nil, operationFault(fs.ErrInvalid), nil
	}
	data := make([]byte, int(size))
	count, err := binding.file.ReadAt(data, offset)
	return data[:count], operationFault(err), nil
}

func (binding *fileBinding) WriteAt(_ context.Context, data []uint8, offset int64) (int64, OsOperationFault, error) {
	binding.mu.Lock()
	defer binding.mu.Unlock()
	if binding.file == nil {
		return 0, operationFault(fs.ErrClosed), nil
	}
	count, err := binding.file.WriteAt(data, offset)
	return int64(count), operationFault(err), nil
}

func (binding *fileBinding) Seek(_ context.Context, offset, whence int64) (int64, OsOperationFault, error) {
	binding.mu.Lock()
	defer binding.mu.Unlock()
	if binding.file == nil {
		return 0, operationFault(fs.ErrClosed), nil
	}
	nativeWhence := int(whence)
	if int64(nativeWhence) != whence {
		return 0, operationFault(fs.ErrInvalid), nil
	}
	position, err := binding.file.Seek(offset, nativeWhence)
	return position, operationFault(err), nil
}

func (binding *fileBinding) Stat(context.Context) (OsFileMetadata, OsOperationFault, error) {
	binding.mu.Lock()
	defer binding.mu.Unlock()
	if binding.file == nil {
		return OsFileMetadata{}, operationFault(fs.ErrClosed), nil
	}
	info, err := binding.file.Stat()
	metadata, fault := fileInfoResult(info, err)
	return metadata, fault, nil
}

func (binding *fileBinding) ReadDir(_ context.Context, count int64) ([]OsFileMetadata, OsOperationFault, error) {
	binding.mu.Lock()
	defer binding.mu.Unlock()
	if binding.file == nil {
		return nil, operationFault(fs.ErrClosed), nil
	}
	if count <= 0 {
		count = 0
	}
	nativeCount := int(count)
	if int64(nativeCount) != count {
		return nil, operationFault(fs.ErrInvalid), nil
	}
	entries, readErr := binding.file.ReadDir(nativeCount)
	metadata, metadataErr := directoryEntries(entries)
	if readErr == nil {
		readErr = metadataErr
	}
	return metadata, operationFault(readErr), nil
}

func (binding *fileBinding) Close(context.Context) error {
	if binding == nil {
		return nil
	}
	binding.mu.Lock()
	defer binding.mu.Unlock()
	if binding.file == nil {
		return nil
	}
	file := binding.file
	binding.file = nil
	return file.Close()
}

func fileInfoResult(info fs.FileInfo, err error) (OsFileMetadata, OsOperationFault) {
	if err != nil || info == nil {
		if err == nil {
			err = errors.New("filesystem returned nil file info")
		}
		return OsFileMetadata{}, operationFault(err)
	}
	return fileInfoValue(info), OsOperationFault{}
}

func directoryEntries(entries []fs.DirEntry) ([]OsFileMetadata, error) {
	values := make([]OsFileMetadata, 0, len(entries))
	for _, entry := range entries {
		info, err := entry.Info()
		if err != nil {
			return values, err
		}
		values = append(values, fileInfoValue(info))
	}
	return values, nil
}

func fileInfoValue(info fs.FileInfo) OsFileMetadata {
	if info == nil {
		return OsFileMetadata{}
	}
	return OsFileMetadata{
		Name: info.Name(), Size: info.Size(), Mode: uint32(info.Mode()),
		ModifiedSeconds: info.ModTime().Unix(), ModifiedNanoseconds: int64(info.ModTime().Nanosecond()),
		Directory: info.IsDir(),
	}
}

func operationFault(err error) OsOperationFault {
	if err == nil {
		return OsOperationFault{}
	}
	code, message := operationError(err)
	return OsOperationFault{Code: code, Message: message}
}

func operationError(err error) (string, string) {
	switch {
	case errors.Is(err, errors.ErrUnsupported):
		return "unsupported", err.Error()
	case errors.Is(err, fs.ErrInvalid):
		return "invalid", err.Error()
	case errors.Is(err, fs.ErrPermission):
		return "permission", err.Error()
	case errors.Is(err, fs.ErrExist):
		return "exist", err.Error()
	case errors.Is(err, fs.ErrNotExist):
		return "not_exist", err.Error()
	case errors.Is(err, fs.ErrClosed):
		return "closed", err.Error()
	case errors.Is(err, io.EOF):
		return "eof", err.Error()
	default:
		return "io", fmt.Sprint(err)
	}
}

File: playground/runtime-rust/src/stdlib_host/filesystem.rs
use super::{BlockingPool, os_binding::*};
use crate::rpc::*;
use std::sync::Arc;

pub type Fault = OsOperationFault;
pub type Metadata = OsFileMetadata;
pub type FsResult<T> = std::result::Result<T, Fault>;
pub fn fault(code: &str) -> Fault {
    Fault {
        code: code.into(),
        message: code.into(),
    }
}
pub fn empty_metadata() -> Metadata {
    Metadata {
        name: String::new(),
        size: 0,
        mode: 0,
        modified_seconds: 0,
        modified_nanoseconds: 0,
        directory: false,
    }
}

pub trait File: Send + Sync {
    fn read(&self, size: i64, offset: Option<i64>) -> (Option<Vec<u8>>, Fault);
    fn write(&self, data: &[u8], offset: Option<i64>) -> (i64, Fault);
    fn seek(&self, offset: i64, whence: i64) -> FsResult<i64>;
    fn stat(&self) -> FsResult<Metadata>;
    fn read_dir(&self, count: i64) -> (Option<Vec<Metadata>>, Fault);
    fn close(&self) -> FsResult<()>;
}
pub trait Filesystem: Send + Sync {
    fn open(&self, name: &str, flags: i64, mode: u32) -> FsResult<Arc<dyn File>>;
    fn stat(&self, name: &str, follow: bool) -> FsResult<Metadata>;
    fn read_dir(&self, name: &str) -> FsResult<Vec<Metadata>>;
    fn mkdir(&self, name: &str, mode: u32) -> FsResult<()>;
    fn remove(&self, name: &str) -> FsResult<()>;
    fn rename(&self, old: &str, new: &str) -> FsResult<()>;
    fn read_link(&self, name: &str) -> FsResult<String>;
    fn chtimes(&self, name: &str, atime: (i64, i64), mtime: (i64, i64)) -> FsResult<()>;
    fn working_directory(&self) -> FsResult<String>;
    fn cache_directory(&self) -> FsResult<String>;
}
pub struct FilesystemProvider {
    pub backend: Arc<dyn Filesystem>,
    pub pool: Arc<BlockingPool>,
}
impl FilesystemProvider {
    pub fn provider(self) -> Result<Arc<dyn Provider>> {
        os_filesystem_provider(Arc::new(self))
    }
}
impl OsFilesystemHandler for FilesystemProvider {
    fn open(
        &self,
        context: CallContext,
        name: String,
        flags: i64,
        mode: u32,
    ) -> BoxFuture<'_, Result<(Option<Arc<dyn OsFileHandleHandler>>, Fault)>> {
        Box::pin(async move {
            context.check()?;
            let backend = self.backend.clone();
            let pool = self.pool.clone();
            self.pool.run(move || match backend.open(&name, flags, mode) {
                Ok(file) => (Some(Arc::new(FileBinding::new(file, pool)) as Arc<dyn OsFileHandleHandler>), fault("")),
                Err(error) => (None, error),
            }).await
        })
    }
    fn stat(
        &self,
        context: CallContext,
        name: String,
        follow: bool,
    ) -> BoxFuture<'_, Result<(Metadata, Fault)>> {
        Box::pin(async move {
            context.check()?;
            let backend = self.backend.clone();
            self.pool
                .run(move || match backend.stat(&name, follow) {
                    Ok(value) => (value, fault("")),
                    Err(error) => (empty_metadata(), error),
                })
                .await
        })
    }
    fn read_dir(
        &self,
        context: CallContext,
        name: String,
    ) -> BoxFuture<'_, Result<(Option<Vec<Metadata>>, Fault)>> {
        Box::pin(async move {
            context.check()?;
            let backend = self.backend.clone();
            self.pool
                .run(move || match backend.read_dir(&name) {
                    Ok(value) => (Some(value), fault("")),
                    Err(error) => (None, error),
                })
                .await
        })
    }
    fn mkdir(
        &self,
        context: CallContext,
        name: String,
        mode: u32,
    ) -> BoxFuture<'_, Result<(Fault,)>> {
        Box::pin(async move {
            context.check()?;
            let backend = self.backend.clone();
            self.pool
                .run(move || {
                    (backend
                        .mkdir(&name, mode)
                        .err()
                        .unwrap_or_else(|| fault("")),)
                })
                .await
        })
    }
    fn remove(&self, context: CallContext, name: String) -> BoxFuture<'_, Result<(Fault,)>> {
        Box::pin(async move {
            context.check()?;
            let backend = self.backend.clone();
            self.pool
                .run(move || (backend.remove(&name).err().unwrap_or_else(|| fault("")),))
                .await
        })
    }
    fn rename(
        &self,
        context: CallContext,
        old: String,
        new: String,
    ) -> BoxFuture<'_, Result<(Fault,)>> {
        Box::pin(async move {
            context.check()?;
            let backend = self.backend.clone();
            self.pool
                .run(move || {
                    (backend
                        .rename(&old, &new)
                        .err()
                        .unwrap_or_else(|| fault("")),)
                })
                .await
        })
    }
    fn readlink(
        &self,
        context: CallContext,
        name: String,
    ) -> BoxFuture<'_, Result<(String, Fault)>> {
        Box::pin(async move {
            context.check()?;
            let backend = self.backend.clone();
            self.pool
                .run(move || match backend.read_link(&name) {
                    Ok(value) => (value, fault("")),
                    Err(error) => (String::new(), error),
                })
                .await
        })
    }
    fn chtimes(
        &self,
        context: CallContext,
        name: String,
        a: i64,
        an: i64,
        m: i64,
        mn: i64,
    ) -> BoxFuture<'_, Result<(Fault,)>> {
        Box::pin(async move {
            context.check()?;
            let backend = self.backend.clone();
            self.pool
                .run(move || {
                    (backend
                        .chtimes(&name, (a, an), (m, mn))
                        .err()
                        .unwrap_or_else(|| fault("")),)
                })
                .await
        })
    }
    fn getwd(&self, context: CallContext) -> BoxFuture<'_, Result<(String, Fault)>> {
        Box::pin(async move {
            context.check()?;
            let backend = self.backend.clone();
            self.pool
                .run(move || match backend.working_directory() {
                    Ok(value) => (value, fault("")),
                    Err(error) => (String::new(), error),
                })
                .await
        })
    }
    fn user_cache_dir(&self, context: CallContext) -> BoxFuture<'_, Result<(String, Fault)>> {
        Box::pin(async move {
            context.check()?;
            let backend = self.backend.clone();
            self.pool
                .run(move || match backend.cache_directory() {
                    Ok(value) => (value, fault("")),
                    Err(error) => (String::new(), error),
                })
                .await
        })
    }
}
struct FileBinding {
    file: Arc<dyn File>,
    pool: Arc<BlockingPool>,
    closing: crate::ffi::Cancellation,
    done: tokio::sync::watch::Receiver<Option<Result<()>>>,
}
impl FileBinding {
    fn new(file: Arc<dyn File>, pool: Arc<BlockingPool>) -> Self {
        let closing = crate::ffi::Cancellation::default();
        let backend = file.clone();
        let done = pool.spawn_cleanup(closing.clone(), move || {
            backend
                .close()
                .map_err(|error| Status::new(error.code, error.message))
        });
        Self {
            file,
            pool,
            closing,
            done,
        }
    }
}
impl Drop for FileBinding {
    fn drop(&mut self) {
        self.closing.cancel();
    }
}
impl OsFileHandleHandler for FileBinding {
    fn read(
        &self,
        context: CallContext,
        size: i64,
    ) -> BoxFuture<'_, Result<(Option<Vec<u8>>, Fault)>> {
        Box::pin(async move {
            context.check()?;
            let file = self.file.clone();
            self.pool
                .run(move || {
                    let (data, mut error) = file.read(size, None);
                    if error.code == "eof" {
                        error = fault("");
                    }
                    (data, error)
                })
                .await
        })
    }
    fn read_at(
        &self,
        context: CallContext,
        size: i64,
        offset: i64,
    ) -> BoxFuture<'_, Result<(Option<Vec<u8>>, Fault)>> {
        Box::pin(async move {
            context.check()?;
            let file = self.file.clone();
            self.pool.run(move || file.read(size, Some(offset))).await
        })
    }
    fn write(
        &self,
        context: CallContext,
        data: Option<Vec<u8>>,
    ) -> BoxFuture<'_, Result<(i64, Fault)>> {
        Box::pin(async move {
            context.check()?;
            let file = self.file.clone();
            self.pool
                .run(move || file.write(&data.unwrap_or_default(), None))
                .await
        })
    }
    fn write_at(
        &self,
        context: CallContext,
        data: Option<Vec<u8>>,
        offset: i64,
    ) -> BoxFuture<'_, Result<(i64, Fault)>> {
        Box::pin(async move {
            context.check()?;
            let file = self.file.clone();
            self.pool
                .run(move || file.write(&data.unwrap_or_default(), Some(offset)))
                .await
        })
    }
    fn seek(
        &self,
        context: CallContext,
        offset: i64,
        whence: i64,
    ) -> BoxFuture<'_, Result<(i64, Fault)>> {
        Box::pin(async move {
            context.check()?;
            let file = self.file.clone();
            self.pool
                .run(move || match file.seek(offset, whence) {
                    Ok(value) => (value, fault("")),
                    Err(error) => (0, error),
                })
                .await
        })
    }
    fn stat(&self, context: CallContext) -> BoxFuture<'_, Result<(Metadata, Fault)>> {
        Box::pin(async move {
            context.check()?;
            let file = self.file.clone();
            self.pool
                .run(move || match file.stat() {
                    Ok(value) => (value, fault("")),
                    Err(error) => (empty_metadata(), error),
                })
                .await
        })
    }
    fn read_dir(
        &self,
        context: CallContext,
        count: i64,
    ) -> BoxFuture<'_, Result<(Option<Vec<Metadata>>, Fault)>> {
        Box::pin(async move {
            context.check()?;
            let file = self.file.clone();
            self.pool.run(move || file.read_dir(count)).await
        })
    }
    fn close(&self) -> BoxFuture<'_, Result<()>> {
        self.closing.cancel();
        Box::pin(async move {
            let mut done = self.done.clone();
            let result = done
                .wait_for(Option::is_some)
                .await
                .map_err(|_| Status::new("internal", "file cleanup owner failed"))?;
            result.as_ref().unwrap().clone()
        })
    }
}
